src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php

```diff
@@ -1863,24 +1863,59 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
         $container->getDefinition('messenger.retry_strategy_locator')
             ->replaceArgument(0, $transportRetryReferences);
 
+        $failureTransports = [];
+        $failureTransportsByTransportName = [];
+
+        $failureTransportsServiceLocatorId = 'messenger.failure_transports.locator';
+        $failureTransportsByTransportNameServiceLocatorId = 'messenger.failure_transports_by_transport_name.locator';
+
         if ($config['failure_transport']) {
             if (!isset($senderReferences[$config['failure_transport']])) {
                 throw new LogicException(sprintf('Invalid Messenger configuration: the failure transport "%s" is not a valid transport or service id.', $config['failure_transport']));
             }
 
-            $container->getDefinition('messenger.failure.send_failed_message_to_failure_transport_listener')
-                ->replaceArgument(0, $senderReferences[$config['failure_transport']]);
+            $failureTransports[$config['failure_transport']] = $senderReferences[$config['failure_transport']];
+        }
+
+        foreach ($config['transports'] as $name => $transport) {
+            if ($transport['failure_transport']) {
+                if (!isset($config['transports'][$transport['failure_transport']])) {
+                    throw new LogicException(sprintf('Invalid Messenger configuration: the failure transport "%s" is not a valid transport or service id.', $transport['failure_transport']));
+                }
+
+                $failureTransportsByTransportName[$name] = $senderReferences[$transport['failure_transport']];
+                $failureTransports[$transport['failure_transport']] = $senderReferences[$transport['failure_transport']];
+            }
+        }
+
+        if (\count($failureTransports) > 0) {
+            $failureTransportsServiceLocator = ServiceLocatorTagPass::register($container, $failureTransports, $failureTransportsServiceLocatorId);
+            $container->getDefinition($failureTransportsServiceLocatorId)
+                ->replaceArgument(0, $failureTransports);
+
+            $globalFailureReceiver = $config['failure_transport'] ?? null;
             $container->getDefinition('console.command.messenger_failed_messages_retry')
-                ->replaceArgument(0, $config['failure_transport']);
+                ->replaceArgument(0, $globalFailureReceiver)
+                ->replaceArgument(1, $senderReferences[$config['failure_transport']] ?? null)
+                ->replaceArgument(5, $container->getDefinition($failureTransportsServiceLocator));
             $container->getDefinition('console.command.messenger_failed_messages_show')
-                ->replaceArgument(0, $config['failure_transport']);
+                ->replaceArgument(0, $globalFailureReceiver)
+                ->replaceArgument(1, $container->getDefinition($failureTransportsServiceLocatorId));
```

Comment:
these should be references, right? We don't normally set a Definition directly on an argument. And should it use `$failureTransportsServiceLocatorId` or `$failureTransportsServiceLocator` - I'm getting lost (this stuff confuses me) in which is which and what should be used.